Muon and AdaMuon: Orthogonal Optimizers for AdvOptim

Koratahiu opened the pull request:
This PR introduces **Muon** (added) and **AdaMuon** (~~to be added~~ added) as new experimental optimizers for `adv-optm`.

1. **[Muon](https://kellerjordan.github.io/posts/muon/)**: An orthogonalizing optimizer that has demonstrated strong performance across diverse tasks. It is competitive with Adam while maintaining only a single momentum state.

2. **[AdaMuon](https://arxiv.org/abs/2507.11005)**: Builds on Muon by incorporating an element-wise second-moment estimator and sign-stabilized orthogonal updates, enabling adaptive scaling while preserving stable update geometry.

---

### Notes

- ~~I implemented the reshaping from SMMF directly in Muon (controlled by `1D Vector Reshape`) to make it work for all dims and shapes. It should be fine since Muon already flattens high-dim params to 2D; reshaping 1D params for the orthogonal update should result in higher quality while ensuring that Muon won't fallback to normal SGD.~~
- Select **5×–20× higher LR than Adam** for Muon.
- For AuxAdam select your usual Adam LR.
- Click the three dots near `MuonWithAuxAdam` to open AdamW_Adv settings that will be used as auxiliary optimizer with Muon.
- `Simplified_AdEMAMix` method proved to work very well with Muon/AdaMuon in my tests, and it keeps the LR range of Muon (no need to decrease the LR). And it's now added as an option for both. You still need to tune beta1 depending on the training length.

---

### Usage
```
git fetch origin pull/1064/head:muon
git checkout muon
venv/scripts/activate
pip install adv-optm==1.2.dev14 --upgrade
```

---

### TODO
- [x] Compare Muon against Adam  
- ~~Test `Factored=True` for Muon~~ Works
- [x] Implement MuonWithAuxAdam
- [x] Add AdaMuon  
- [x] Compare AdaMuon against Muon~~
- ~~Test `Factored=True` for AdaMuon~~ Works
- ~~Maybe add as an option to `Prodigy_adv` to test if it works with Prodigy~~ (not working).
- [x] Enhance the auto preset for MuonWithAuxAdam and make it model-specific 

---

dxqb commented:
> Why layer filter, though? If it’s because of 1D params, we already reshape them to 2D effectively by the SMMF method (when `1D Vector Reshape` = True).

I cannot argue the theory, but you'll find somewhere in the repo of the original author why they have introduced the AuxAdam variant and why they think Muon cannot work well on even some layers with 2D parameters.

Koratahiu commented:
Added MuonWithAuxAdam optimizer to TODO list

Koratahiu commented:
> would be great if you could add the original Muon as well not only Muon_Adv, so we have a comparison to the original

Yeah, I meant as an option to Muon_Adv, as it will be reduced to the original when you disable `1D Vector Reshape`.

> a layer filter that cannot be automated but the user has to choose

Why is that? I am thinking of adding an option to automate it, since the models we support are limited, and it should be easy to implement.
Also, most users won't bother looking at the architecture of the models and extracting the names of non-hidden layers. And those who want to manually choose can disable it and provide their own layer keys.
I prefer this approach

dxqb commented on modules/util/build_muon_adam_key_fn.py:
```diff
@@ -0,0 +1,207 @@
+                ]
+            case ModelType.FLUX_DEV_1 | ModelType.CHROMA_1:
+                default_patterns = [
+                    'context_embedder',
```
at least for Flux, these layers are all layers except the transformer blocks. Is this the case for all (or many) models?
Wouldn't it be easier to define the hidden-layers then?

This is now a re-usable component in this PR: https://github.com/Nerogar/OneTrainer/pull/1091
<img width="406" height="106" alt="image" src="https://github.com/user-attachments/assets/153bc3a8-5bae-4498-af5f-95ad55711de5" />

Koratahiu replied:
Done

Do the hidden layers provided by users have any use? I think they're now redundant since we filter all hidden layers for currently supported models.

dxqb commented on modules/util/build_muon_adam_key_fn.py:
```diff
@@ -0,0 +1,207 @@
+                    'img_in',
+                    'txt_in',
+                    'txt_norm',
+                    'final_layer_norm',
```
these layers never exist. they are keys of the original Flux inference code. even if we load such a file through diffusers, we end up with diffusers keys.

dxqb reviewed (commented):
the code is nicely isolated now.
the Muon-specific changes now are, roughly summarized:

- [x]  - UI code (ok)
- [ ]  - code in create_optimizer (ok)
- [x]  - tensorboard reporting code (ok)
- [ ]  - backup code: I plan to have a second look here if this can be avoided. I'm hesitant to change backup code, which could introduce subtle regressions?
- [ ]  - I'll probably want to do at least a prototype/draft implementation of the original Muon repo, to make sure it can be integrated in the same way

Koratahiu commented:
- [ ] In [PyTorch 2.9, Muon](https://docs.pytorch.org/docs/stable/generated/torch.optim.Muon.html#torch.optim.Muon) uses RMS scaling, which scales its updates to match Adam’s learning rate range.  
  In this PR, we already have this feature, but it’s only enabled for NorMuon/AdaMuon, so change this behaviour and add it as a new boolean flag, defaulting to `True`.

dxqb commented on modules/ui/OptimizerParamsWindow.py:
```diff
@@ -177,6 +180,20 @@ def create_dynamic_ui(
             'k_warmup_steps': {'title': 'K-β Warmup Steps ', 'tooltip': 'When using Kourkoutas Beta, the number of initial training steps during which the dynamic β₂ logic is held off. In this period, β₂ is set to its fixed value to allow for initial training stability before the adaptive mechanism activates.', 'type': 'int'},
             'schedulefree_c': {'title': 'Schedule free averaging strength', 'tooltip': 'Larger values = more responsive (shorter averaging window); smaller values = smoother (longer window). Set to 0 to disable and use the original Schedule-Free rule. Short small batches (≈6-12); long/large-batch (≈50-200).', 'type': 'float'},
+            'ns_steps': {'title': 'Newton-Schulz Iterations', 'tooltip': 'Controls the number of iterations for update orthogonalization. Higher values improve the updates quality but make each step slower. Lower values are faster per step but may be less effective.', 'type': 'int'},
+            'MuonWithAuxAdam': {'title': 'MuonWithAuxAdam', 'tooltip': 'whatever to use the standard way of Muon, non hidden layers fallback to ADAMW_ADV and MUON_ADV takes the rest.', 'type': 'bool'},
```
tooltip: I don't think you start a sentence with "whatever" in this case

dxqb commented on modules/util/create.py:
```diff
@@ -1172,6 +1174,97 @@ def create_optimizer(
+                rms_rescaling=optimizer_config.rms_rescaling if optimizer_config.rms_rescaling is not None else True,
+                nnmf_factor=optimizer_config.nnmf_factor if optimizer_config.nnmf_factor is not None else False,
+                stochastic_rounding=optimizer_config.stochastic_rounding,
+                nesterov=optimizer_config.nesterov if optimizer_config.nesterov is not None else True,
```
what is this and why is it enabled?
The (previously existing) tooltip doesn't explain anything.

Explain all options perfectly for advanced optimizers is probably too much to expect from tooltips, but an option that is enabled by default I'd like to understand as a user.

Koratahiu replied:
The Muon author ran some experiments and found that Nesterov momentum is always better than normal momentum for Muon in all tests, so he hardcoded it to `True` in the original Muon (you can’t disable it there unless you edit the code).

I set it to `True` by default to follow his choice, but here you can disable it to use normal momentum, or even use AdEMAMix momentum.

As for the tooltip, it’s shared with the older optimizers, so I left it unchanged.

We have the wiki for these things I guess

dxqb commented on modules/util/build_muon_adam_key_fn.py:
```diff
@@ -0,0 +1,196 @@
+                ]
+            case ModelType.STABLE_DIFFUSION_3 | ModelType.STABLE_DIFFUSION_35 | ModelType.SANA  | ModelType.FLUX_DEV_1  | ModelType.CHROMA_1  | ModelType.QWEN  | ModelType.PIXART_ALPHA | ModelType.PIXART_SIGMA:
+                default_patterns = [
+                    'transformer_blocks',
```
as I understand these filters, AuxAdam will be irrelevant for almost all LoRA training use cases, unless you train a "full" Lora. I think I'd like to know this as user. Can we explain that at the "MuonWithAuxAdam" option? Otherwise I might spend much time finetuning the Adam parameters and it isn't even used

Koratahiu replied:
Added a note for this in tooltip

dxqb commented on modules/ui/OptimizerParamsWindow.py:
```diff
@@ -177,6 +180,20 @@ def create_dynamic_ui(
+            'muon_adam_lr': {'title': 'Auxiliary Adam LR', 'tooltip': 'Learning rate for the auxiliary AdamW optimizer. If empty, it will use the main learning rate.', 'type': 'float'},
+            'muon_te1_adam_lr': {'title': 'AuxAdam TE1 LR', 'tooltip': 'Learning rate for the auxiliary AdamW optimizer for the first text encoder. If empty, it will use the Auxiliary Adam LR.', 'type': 'float'},
+            'muon_te2_adam_lr': {'title': 'AuxAdam TE2 LR', 'tooltip': 'Learning rate for the auxiliary AdamW optimizer for the second text encoder. If empty, it will use the Auxiliary Adam LR.', 'type': 'float'},
+            'rms_rescaling': {'title': 'RMS Rescaling', 'tooltip': 'RMS-aligned rescaling for the optimizer update. Allows for the reuse of existing Adam learning rate schedules.', 'type': 'bool'},
```
This is enabled by default so I'd like to understand it as a user. Does the tooltip mean I can use cosine schedulers with that, but not without this? Why not?
Or does this mean that it can use similar learning rates as Adam then?

Koratahiu replied:
The same learning rate as Adam.  
I’ll reword this to be clearer, and note that it also makes the optimizer use the same weight decay as Adam, since weight decay is scaled by the learning rate.
I might mention this too

dxqb commented on modules/util/build_muon_adam_key_fn.py:
```diff
@@ -0,0 +1,196 @@
+                    param_map[id(p)] = get_optim_type(full_param_name, p)
+                    all_processed_params.append(p)
+
+    # Print a summary for verification
```
`WARNING: The following hidden layer patterns did not match any parameters: ['encoder.block']`    

Is that okay? Standard LoRA with Chroma

Koratahiu replied:
Yeah, that's the hidden layer of T5
I will remove this warning as it creates confusion in these scenarios

Koratahiu commented on modules/util/ui/UIState.py:
```diff
@@ -168,6 +168,9 @@ def update(_0, _1, _2):
     def __create_vars(self, obj):
         new_vars = {}
 
+        if obj is None:
```
Yeah, it’s a “bandaid fix.”  
The Muon window defaults to `None`.  
Can you suggest an alternative if this is too fragile?

O-J1 commented on modules/util/ui/UIState.py:
```diff
@@ -241,12 +244,17 @@ def __create_vars(self, obj):
 
         if is_config:
             for name, var_type in obj.types.items():
+                if name not in self.__vars:
```
Same for both 256-257 and 276-277, set_vars is meant to update existing UI variables, Im fairly sure this would silently skip them.

What scenario necessitated you to add this? I havent seen errors with set vars so far on `main`

dxqb commented on modules/modelSetup/BaseModelSetup.py:
```diff
@@ -106,12 +106,26 @@ def report_to_tensorboard(
-            if name not in reported_learning_rates:
-                reported_learning_rates[name] = lr
+        # Handle MuonWithAuxAdam's split parameter groups
+        if any('optim_type' in g for g in model.optimizer.param_groups):
```
does this work with original muon? I guess no because original muon differentiates by setting `param_group['use_muon']`

dxqb commented on modules/util/build_muon_adam_key_fn.py:
```diff
@@ -0,0 +1,199 @@
+            adam_group['optim_type'] = 'adam'
+            # Set Adam-specific LR
+            base_adam_lr = optimizer_config.muon_adam_lr if optimizer_config.muon_adam_lr is not None else config.learning_rate
+            te1_adam_lr = optimizer_config.muon_te1_adam_lr
```
I don't very much like having specialized code for TE1 and TE2, and prefer not support training text encoders with Muon/AuxAdam. But I guess the use case for SD and SDXL is important enough to justify this?

Koratahiu commented on modules/util/build_muon_adam_key_fn.py:
```diff
@@ -0,0 +1,199 @@
+            case _: # Unmatched cases
+                raise NotImplementedError(f"Default hidden layer patterns are not defined for model type: {model.model_type}")
+        filters = [ModuleFilter(p, use_regex=False) for p in default_patterns]
+        if True:
```
Yeah, but then it was too useful to be (if config.debug). I guess we have three options:

1. Remove (if True) and print the summary for the users
2. If config.debug
3. a warning when using AuxAdam, but it's 100% Muon

Koratahiu commented on modules/util/build_muon_adam_key_fn.py:
```diff
@@ -0,0 +1,199 @@
+        'unet_lora', 'transformer_lora'
+    ]
+
+    for module_prefix in sub_modules_to_check:
```
I thought of this, the list should be somewhere else

dxqb replied:
this is still open

dxqb commented on modules/util/optimizer_util.py:
```diff
@@ -529,6 +548,59 @@ def init_model_parameters(
         "cautious_mask": False,
         "orthogonal_gradient": False,
     },
+    Optimizer.MUON_ADV: {
```
does this just distribute Muon-workload onto multiple GPUs, or is the optimizer incorrect if SingleDeviceMuon is ran on multiple GPUs?
As long as SingleDeviceMuon is deterministic, I don't see why it would fail. If so, we can go ahead with the PR, because then multi-GPU support would only be an optimization.

Koratahiu replied:
> does this just distribute Muon-workload onto multiple GPUs, or is the optimizer incorrect if SingleDeviceMuon is ran on multiple GPUs? As long as SingleDeviceMuon is deterministic, I don't see why it would fail. If so, we can go ahead with the PR, because then multi-GPU support would only be an optimization.

It is primarily for performance; instead of computing NS N times (where N is the number of GPUs), it computes it once per layer between the GPUs.
I implemented it anyway, but I still don't think it will work with fused backpass.

Koratahiu commented on modules/util/build_muon_adam_key_fn.py:
```diff
@@ -0,0 +1,199 @@
+    def layer_key_fn(p: torch.nn.Parameter) -> str:
+        return param_map.get(id(p), 'adam')
+
+    return layer_key_fn
```
It's a left-over and removed

Koratahiu commented on modules/util/optimizer/muon_util.py:
```diff
@@ -0,0 +1,199 @@
+from collections.abc import Callable
```
done

Koratahiu commented:
I integrated the latest AdvOptm version in this PR. This version (1.2.8) includes:

- [Cautious Weight Decay](https://arxiv.org/abs/2510.12402) for all adv optimizers (a toggle feature that I observe is getting significant adoption in ML repositories).
- Improved parameter update and weight decay for BF16 SR. The updates are now accumulated in float32 and rounded once at the end.
- Fixed several bugs in Cautious and Grams options. I remember @miasik was getting NaNs, this might be the cause. 
- Implemented the original Muon logic for multi-GPUs in Muon adv variants. This logic should be faster and more accurate (for multi-GPUs), but I suspect it is incompatible with fused backpass?

dxqb commented on modules/util/optimizer/muon_util.py:
```diff
@@ -133,10 +133,7 @@ def get_optim_type(param_name: str, p: torch.nn.Parameter) -> str:
-        return param_map.get(id(p), 'adam')
-
-    return layer_key_fn
+    return lambda p: param_map.get(id(p), 'adam')
```
yes

Koratahiu commented on modules/util/optimizer/muon_util.py:
```diff
@@ -0,0 +1,173 @@
+        # Create filters from the user's configuration string
+        patterns_list = [p.strip() for p in config.optimizer.muon_hidden_layers.split(',') if p.strip()]
+        filters = [ModuleFilter(p, use_regex=config.optimizer.muon_adam_regex) for p in patterns_list]
+        if True:
```
Sorry, I forgot to remove this
